iceoryx_examples/icediscovery/iox_wait_for_service.cpp

```diff
@@ -44,14 +46,15 @@ void printSearchResult(const discovery::ServiceContainer& result)
     }
 }
 
-Discovery* discoveryPtr{nullptr};
+volatile Discovery* discoverySigHandlerAccess{nullptr};
 
 void sigHandler(int)
 {
     //! [unblock wait]
-    if (discoveryPtr)
+    keepRunning = false;
```

Comment:
Why do we need `keepRunning` here? Shouldn't `discoverySigHandlerAccess` suffice?

Reply:
They have two different purposes. `discoverySigHandlerAccess` could be set to a `nullptr` to indicate the while-loop to end. But then `discoverySigHandlerAccess` would be misused to indicate the loop condition and it would be less clear than with `keepRunning`. In general I try to avoid surrogates to not frustrate my future self too much.

Reply:
Okay, I got it, but I have another question: shouldn't we use an atomic volatile bool to ensure that `keepRunning` is synchronized?

Reply:
The variable does not need to be thread safe but only async signal safe, i.e. just atomic with respect to signals. This means that read/write operations on variables must be executed in one instruction. Furthermore, `volatile` tells the compiler that the variable can be changed from another location not visible to the compiler and therefore it always needs to be read from the memory and not from the register and also cannot be optimized away. Keep in mind, the OS does not start a new thread for the execution of the signal handler but executes it in an already [running thread](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xsh_chap02.html#tag_22_02_04_02). You can kind of think of it like a `goto` to the signal handler when the signal is raised. This is also the reason why e.g. `pthread_mutex_lock` is not [signal safe](https://man7.org/linux/man-pages/man7/signal-safety.7.html) since there might be nobody left to unlock the mutex if it is already locked.

In [many](https://en.cppreference.com/w/c/program/sig_atomic_t) [examples](https://www.gnu.org/software/libc/manual/html_node/Handler-Returns.html) for the signal handler one can find a definition like `volatile sig_atomic_t keepRunning = 1;`. In this case `sig_atomic_t` is not an atomic data type but just a type alias to an integer type which can be accessed as an [atomic](https://en.cppreference.com/w/c/program/sig_atomic_t) [entity](https://www.gnu.org/software/libc/manual/html_node/Atomic-Types.html). This is also true for pointer and since a boolean is smaller than an `int`, it is also true for `bool`.

[Here](https://www.gnu.org/software/libc/manual/html_node/Non_002datomic-Example.html) you can find a good explanation about non-atomic access. Basically it boils down to torn writes and torn reads.

If you are more interested in the topic and have some time, this is an interesting read -> https://www.gnu.org/software/libc/manual/html_node/Signal-Handling.html

Reply:
Btw, here is a short example https://godbolt.org/z/deqsPjcsY
```cpp
volatile int v = 0;             // (1)
volatile sig_atomic_t va = 0;   // (2)
std::atomic<int> a = 0;         // (3)

struct Point {
    uint64_t x {0};
    uint64_t y {0};
} p;                            // (4)

void setVolatile(int num) {
    v = num;
}

void setVolatileSigAtomic(int num) {
    va = num;
}

void setAtomic(int num) {
    a.store(num, std::memory_order_relaxed);
}

void setPoint(Point point) {
    p = point;
}
```
When compiled with `-O1` this is the result
```asm
setVolatile(int):
        mov     DWORD PTR v[rip], edi
        ret
setVolatileSigAtomic(int):
        mov     DWORD PTR va[rip], edi
        ret
setAtomic(int):
        mov     DWORD PTR a[rip], edi
        ret
setPoint(Point):
        mov     QWORD PTR p[rip], rdi
        mov     QWORD PTR p[rip+8], rsi
        ret
```
As you can see on , the first three variants compile to the same instruction and are equivalent. The last one does not fulfill the requirement that the operation is atomic. Contrary to atomic data types, the usage of the first two variants is limited since they do not prevent reordering and don't have CAS operations but for our purpose it is equivalent to an atomic which supports only relaxed read and write operations.

Reply:
Ah okay, thanks for the explanation and the links you provided!